src/java.base/unix/classes/sun/nio/fs/UnixPath.java

```diff
@@ -857,13 +860,83 @@ public Path toRealPath(LinkOption... options) throws IOException {
             result = result.resolve(element);
         }
 
-        // check file exists (without following links)
+        // check whether file exists (without following links)
         try {
             UnixFileAttributes.get(result, false);
         } catch (UnixException x) {
             x.rethrowAsIOException(result);
         }
-        return result;
+
+        // Return if the file system is not both case insensitive and retentive
+        if (!fs.isCaseInsensitiveAndRetentive())
+            return result;
+
+        UnixPath path = fs.rootDirectory();
+
+        SecurityManager sm = System.getSecurityManager();
+
+        // Traverse the result obtained above from the root downward, leaving
+        // any '..' elements intact, and replacing other elements with the
+        // entry in the same directory which has an equal key
+        for (int i = 0; i < result.getNameCount(); i++ ) {
+            UnixPath elt = result.getName(i);
+
+            // If the element is "..", append it directly and continue
+            if (elt.toString().equals("..")) {
+                path = path.resolve(elt);
+                continue;
+            }
+
+            // Derive full path to element and check readability
+            UnixPath eltPath = path.resolve(elt);
+            if (sm != null)
+                sm.checkRead(eltPath.getPathForPermissionCheck());
+
+            // Derive element key
+            UnixFileAttributes attrs = null;
+            try {
+                attrs = UnixFileAttributes.get(eltPath, false);
+            } catch (UnixException x) {
+                x.rethrowAsIOException(result);
+            }
+            final UnixFileKey eltKey = attrs.fileKey();
+
+            // Check readbility of path thus far
+            if (sm != null)
+                sm.checkRead(path.getPathForPermissionCheck());
```

Comment:
I don't think this permission check is needed here. The "checkRead" at line 816 does the require check for the method.

Reply:
Removed in 49d18c6a39d9d70b7da91c5b4b3740ba3c890948.